File: resident_app/lib/src/models/community_payment_config.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class DirectUpiConfig {
  const DirectUpiConfig._({required this.enabled, this.vpa, this.payeeName});

  const DirectUpiConfig.disabled() : this._(enabled: false);

  final bool enabled;
  final String? vpa;
  final String? payeeName;

  bool get isUsable =>
      enabled &&
      vpa != null &&
      vpa!.isNotEmpty &&
      payeeName != null &&
      payeeName!.isNotEmpty;

  factory DirectUpiConfig.fromMap(Map<String, dynamic> data) {
    final enabled = data['enabled'];
    if (enabled is! bool) {
      throw const FormatException(
        'Community payment configuration is invalid.',
      );
    }
    if (!enabled) return const DirectUpiConfig.disabled();

    final rawVpa = data['vpa'];
    final rawPayeeName = data['payeeName'];
    if (rawVpa is! String || rawPayeeName is! String) {
      throw const FormatException(
        'Community payment configuration is invalid.',
      );
    }
    final vpa = rawVpa.trim();
    final payeeName = rawPayeeName.trim();
    final parts = vpa.split('@');
    if (vpa.isEmpty ||
        vpa.contains(RegExp(r'\s')) ||
        parts.length != 2 ||
        parts.any((part) => part.isEmpty) ||
        payeeName.isEmpty) {
      throw const FormatException(
        'Community payment configuration is invalid.',
      );
    }
    return DirectUpiConfig._(enabled: true, vpa: vpa, payeeName: payeeName);
  }
}

class CommunityPaymentConfig {
  const CommunityPaymentConfig({
    required this.communityId,
    required this.version,
    required this.directUpi,
    required this.configured,
    this.updatedBy,
    this.updatedAt,
  });

  factory CommunityPaymentConfig.unconfigured(String communityId) {
    final id = communityId.trim();
    if (id.isEmpty) {
      throw ArgumentError('Community ID is required.');
    }
    return CommunityPaymentConfig(
      communityId: id,
      version: 1,
      directUpi: const DirectUpiConfig.disabled(),
      configured: false,
    );
  }

  final String communityId;
  final int version;
  final DirectUpiConfig directUpi;
  final String? updatedBy;
  final DateTime? updatedAt;
  final bool configured;

  factory CommunityPaymentConfig.fromMap(
    String documentCommunityId,
    Map<String, dynamic> data,
  ) {
    final id = documentCommunityId.trim();
    if (id.isEmpty ||
        data['communityId'] != id ||
        data['version'] is! int ||
        data['version'] != 1) {
      throw const FormatException(
        'Community payment configuration is invalid.',
      );
    }

    final rawDirectUpi = data['directUpi'];
    if (rawDirectUpi is! Map ||
        rawDirectUpi.keys.any((key) => key is! String)) {
      throw const FormatException(
        'Community payment configuration is invalid.',
      );
    }

    final updatedBy = data['updatedBy'];
    final rawUpdatedAt = data['updatedAt'];
    final updatedAt = rawUpdatedAt is Timestamp
        ? rawUpdatedAt.toDate()
        : rawUpdatedAt is DateTime
        ? rawUpdatedAt
        : null;
    if (updatedBy is! String || updatedBy.trim().isEmpty || updatedAt == null) {
      throw const FormatException(
        'Community payment configuration is invalid.',
      );
    }

    return CommunityPaymentConfig(
      communityId: id,
      version: 1,
      directUpi: DirectUpiConfig.fromMap(
        Map<String, dynamic>.from(rawDirectUpi),
      ),
      updatedBy: updatedBy.trim(),
      updatedAt: updatedAt,
      configured: true,
    );
  }
}

File: functions/src/community_payment_config.js
const {FieldValue} = require("firebase-admin/firestore");
const {RegistrationError} = require("./register_resident");
const {requireAuthorizedCommunityAdmin} = require("./community_location_management");
const {buildAuditLog} = require("./audit_log");

function invalid(message) {
  throw new RegistrationError("invalid-argument", message);
}

function validateInput(data) {
  if (!data || typeof data !== "object" || Array.isArray(data) ||
      Object.keys(data).some((key) => !["communityId", "directUpi"].includes(key))) {
    invalid("Only communityId and directUpi may be supplied.");
  }
  if (typeof data.communityId !== "string") invalid("A valid community ID is required.");
  const communityId = data.communityId.trim();
  if (!communityId || communityId !== data.communityId || communityId.includes("/") || communityId.length > 128) {
    invalid("A valid community ID is required.");
  }

  const directUpi = data.directUpi;
  if (!directUpi || typeof directUpi !== "object" || Array.isArray(directUpi) ||
      Object.keys(directUpi).some((key) => !["enabled", "vpa", "payeeName"].includes(key)) ||
      typeof directUpi.enabled !== "boolean") {
    invalid("A valid Direct UPI configuration is required.");
  }

  if (!directUpi.enabled) {
    if (Object.keys(directUpi).some((key) => key !== "enabled")) {
      invalid("VPA and payeeName must be omitted when Direct UPI is disabled.");
    }
    return {communityId, directUpi: {enabled: false}};
  }

  if (typeof directUpi.vpa !== "string" || typeof directUpi.payeeName !== "string") {
    invalid("A VPA and payee name are required when Direct UPI is enabled.");
  }
  const vpa = directUpi.vpa.trim();
  const payeeName = directUpi.payeeName.trim();
  const parts = vpa.split("@");
  const vpaPartPattern = /^[A-Za-z0-9._+-]+$/;
  const handlePattern = /^[A-Za-z0-9._-]+$/;
  if (vpa.length > 255 || /\s/.test(vpa) || parts.length !== 2 ||
      !vpaPartPattern.test(parts[0]) || !handlePattern.test(parts[1])) {
    invalid("Enter a valid UPI VPA.");
  }
  if (!payeeName || payeeName.length > 120) invalid("Enter a valid payee name.");

  return {communityId, directUpi: {enabled: true, vpa, payeeName}};
}

async function updateCommunityPaymentConfigCore({db, auth, data}) {
  const input = validateInput(data);
  const {uid} = await requireAuthorizedCommunityAdmin(db, auth, input.communityId);
  const communityRef = db.collection("communities").doc(input.communityId);
  const configRef = db.collection("communityPaymentConfigs").doc(input.communityId);
  const auditRef = db.collection("auditLogs").doc();

  await db.runTransaction(async (transaction) => {
    const communitySnapshot = await transaction.get(communityRef);
    if (!communitySnapshot.exists || communitySnapshot.data()?.isActive !== true) {
      throw new RegistrationError("failed-precondition", "The target community is inactive or unavailable.");
    }

    transaction.set(configRef, {
      communityId: input.communityId,
      version: 1,
      directUpi: input.directUpi,
      updatedBy: uid,
      updatedAt: FieldValue.serverTimestamp(),
    });
    transaction.create(auditRef, buildAuditLog({
      actorUid: uid,
      actorRole: "admin",
      communityId: input.communityId,
      action: "community.payment_config_update",
      targetType: "community_payment_config",
      targetId: input.communityId,
      summary: "Community payment configuration updated.",
      metadata: {directUpiEnabled: input.directUpi.enabled},
    }));
  });

  return {communityId: input.communityId, directUpi: input.directUpi, updatedBy: uid};
}

module.exports = {updateCommunityPaymentConfigCore, validateInput};
